Add "continue as user" option to login.

sgomes opened the pull request:
This PR adds a link for the user to continue as the currently logged-in user as part of the login screen. This is only displayed if the user is already logged in, of course. It's hopefully a clearer mechanism to indicate that the user is already logged in and they don't need to enter their details again.

![image](https://user-images.githubusercontent.com/409615/62880253-e5991d00-bd24-11e9-8f44-87235143847c.png)

In addition, this PR changes the login pages to always use the logged-out layout, in order for the change to make sense as a whole.

This PR started out as a part of #33642 with changes initially proposed by @jsnajdr.

#### Changes proposed in this Pull Request

* Add "continue as user" option to login.
* Force the login pages to use logged-out layout.

#### Testing instructions

* Ensure you're logged in
* Open `/log-in` on the live branch
* Ensure that you see the "continue as user option and that it works normally

In addition to this, please ensure that all other login functionality continues working normally.

sgomes committed:
Add "continue as user" option to login.

This displays a link for the user to continue as the currently logged-in
user as part of the login screen. Only displayed if the user is already
logged in, of course.

This change was originally part of #33642.

sgomes committed:
Force logged-out layout for login pages.

sgomes commented on client/blocks/login/index.jsx:
```diff
@@ -168,12 +169,18 @@ class Login extends Component {
 		let postHeader = null;
 
+		if ( currentUser ) {
+			postHeader = <ContinueAsUser user={ currentUser } redirectUrl={ redirectTo || '/' } />;
```
@jsnajdr : I'm not confident this is the correct way to handle the redirect URL. As far as I can tell, `getRedirectToSanitized` only gets populated **after** a successful login attempt. Is there a better way of getting a sanitised redirect URL here?

jsnajdr replied:
Yes, apparently the validation is server-side and part of the login endpoint. We'll need a separate redirect validation endpoint to make this feature work.

Some pointers I found:
- the server-side validation is done in `Login_Base_Endpoint::create_response_with_redirect_url` method in the wpcom repo.
- was implemented by @stephanethomas and @Tug in D5959-code
- #14872 has interesting historical info how we used to do the validation client-side

Possible solution: new endpoint:
```
/rest/v1.1/users/$user/validate-redirect
```
(modeled after existing `/users/$user/auth-options` endpoint queried after you enter username in login)

One possible caveat: the `login_redirect` and `logout_redirect` filters take `$user` as a parameter, but here they would be called anonymously, before login, being exposed to public. Is it safe to call them in that case? Can they be used to leak info about the user?

@josephscott might help a lot here, knowing about all flavors of WordPress login 🙂

josephscott replied:
I'd suggest any redirect code get some help from the security folks.  We want to be careful not to send people some where outside of our control.

jsnajdr replied:
> One possible caveat: the `login_redirect` and `logout_redirect` filters take `$user` as a parameter, but here they would be called anonymously, before login, being exposed to public.

I realized this is not true: the `/users/$user/validate-redirect` endpoint can be authenticated only. Calypso Login page first realizes that the user is already logged in, and only then offers "continue as user" and verifies the redirect link.

sgomes replied:
Thanks, @jsnajdr and @josephscott! Any suggestions of where I should go for help in making this API endpoint? My PHP-fu is weak 😞

jsnajdr replied:
I'll try to help with the endpoint.

sgomes replied:
Thank yo so much, @jsnajdr! 🙇

jsnajdr replied:
Added new endpoint in D31621:
```
/rest/v1.1/me/validate-redirect?redirect_url=...
```
@sgomes With that patch applied, you can continue developing this feature.

sgomes replied:
Thank you so much for this again, @jsnajdr 🙏 You rock! ❤️

sgomes commented:
@jsnajdr and others: Please feel free to any any folks that you think should be taking a look at this PR. Not having done a lot of feature work in the past, I'm not sure who to loop in. Thanks! 🙂

sgomes commented:
The live branch doesn't appear to be working, as opposed to my development copy. Investigating.

jsnajdr committed:
Export makeForcedLoggedOutLayout from server-side controller, too

jsnajdr commented:
@sgomes I was able to reproduce that crash easily, just by running `npm start` and navigating to `/log-in`. The `makeForcedLoggedOutLayout` function needs to be exported from the server-side `controller/index.node.js`, too. Not only from `controller/index.web.js`. Pushed a fix in f4dd05e.

BTW I'd call the function just `makeLoggedOutLayout` 🙂

sgomes commented:
> @sgomes I was able to reproduce that crash easily, just by running `npm start` and navigating to `/log-in`. The `makeForcedLoggedOutLayout` function needs to be exported from the server-side `controller/index.node.js`, too. Not only from `controller/index.web.js`. Pushed a fix in [f4dd05e](https://github.com/Automattic/wp-calypso/commit/f4dd05e69464d34efb551dec9c2b7cde7d1460c1).
> 
> BTW I'd call the function just `makeLoggedOutLayout` 🙂

Thanks, @jsnajdr, this is exactly it! 🙂 

I think I figured out what happened. Live reloading only works for the client side, so I was still running an unmodified copy of the server side. That's why I wasn't getting any errors.

sgomes committed:
Rename makeForcedLoggedOutLayout.

Renamed to makeLoggedOutLayout.

sgomes commented:
This PR is blocked while waiting for D31621.

sgomes committed:
Add support for redirect validation API endpoint.

sgomes commented:
Added support for the redirect validation API endpoint being introduced in D31621.

It should always use `/` for now, while the endpoint is unavailable, but should start using validated URLs once the change gets merged and the PR is made available.

sgomes commented:
This is no longer blocked, as @jsnajdr's new API endpoint is now in production.

This appears to be working well in the live branch. Ready for review!

jsnajdr commented on client/blocks/login/index.jsx:
```diff
@@ -170,12 +204,25 @@ class Login extends Component {
+			// like that, but it is better than the alternative, and in practice it should happen quicker than
+			// the user can notice.
+			postHeader = (
+				<ContinueAsUser user={ currentUser } redirectUrl={ validatedRedirectToQueryParam || '/' } />
```
`ContinueAsUser` could retrieve the `currentUser` from Redux itself (it's a block after all). And would be also a nicer place for the validation logic. It could also be in some "disabled but rendered" state while validation is in progress. In the end, we could just render:
```jsx
<ContinueAsUser />
```
and the component does all the heavy lifting internally.

sgomes replied:
That's a very good point, and I should have thought of that 😞 

It's done! I opted for leaving `currentUser` in the login block, though, as it makes more sense, semantically speaking, to check for a current user before adding a `ContinueAsUser` block to the output. Let me know if you disagree.

sgomes committed:
Move redirect validation inside `ContinueAsUser`.

jsnajdr commented on client/blocks/login/continue-as-user.jsx:
```diff
@@ -0,0 +1,85 @@
+			// const query = new URLSearchParams( window.location.search );
+			// const redirectUrl = query.get( 'redirect_to' );
+			const query = qsParse( window.location.search, { ignoreQueryPrefix: true } );
+			const redirectUrl = query.redirect_to;
```
I'm not sure if my comment got lost somewhere, so I'll repeat it: the query parameters are in Redux: `state.ui.route.query.current` and can be retrieved with the `getCurrentQueryArguments` selector.

I'm not sure what the `ignoreQueryPrefix` arg does though.

sgomes replied:
I don't know if I missed an earlier comment of yours (sorry if that's the case!), but this is the first time I'm reading this. That's very handy, thank you! Done, and I also simplified and clarified the effect logic a bit.

The `ignoreQueryPrefix` option is just a `qs` thing; all it does is ignore the leading `?` in the query string if it's there. Glad to not have to use it 🙂

jsnajdr replied:
> I don't know if I missed an earlier comment of yours (sorry if that's the case!)

I couldn't find it either when searching for it later, so it probably ended up in an unsubmitted `textarea` in a closed tab 🙂

jsnajdr commented on client/blocks/login/continue-as-user.jsx:
```diff
@@ -0,0 +1,85 @@
+
+function ContinueAsUser( { currentUser } ) {
+	const translate = useTranslate();
+	const [ validated, setValidated ] = useState( null );
```
Can you use `/` as the initial value? It's later defaulted to `/` anyway...

sgomes replied:
Done.

sgomes committed:
Read current query from Redux.

Also simplifies and clarifies effect logic.

jsnajdr committed:
Get redirectUrl from query as a prop instead of passing whole query

Also rename state variable to `validatedRedirectUrl`

jsnajdr committed:
Send redirect_url using a query param, escapes things correctly

jsnajdr committed:
Remove check for currentUser, is guaranteed to be valid

jsnajdr committed:
Extract validateUrl to separate function

Makes it similar to a Redux action and simplifies the effect hook call.

jsnajdr committed:
Add MomentProvider to ReduxWrappedLoggedOutLayout, fixes 2FA screens

jsnajdr committed:
Show ContinueAsUser only on the initial, plain login page

Don't show it on other screens, like 2FA, or on customized logins to Jetpack,
WooCommerce, or other OAuth-connected apps like Crowdsignal.

sgomes commented on client/blocks/login/continue-as-user.jsx:
```diff
@@ -0,0 +1,77 @@
+
+export default connect( state => ( {
+	currentUser: getCurrentUser( state ),
+	redirectUrlFromQuery: get( getCurrentQueryArguments( state ), 'redirect_to', null ),
```
Can't wait until we can use the optional chaining operator (`getCurrentQueryArguments( state )?.redirect_to`)... Using `_.get` for this is a bit like using a sledgehammer to drive a nail into a wall... 🙁

jsnajdr replied:
Yeah, can't wait for optional chaining and friends to become available!

jsnajdr commented:
@sgomes I pushed a few more fixes and cleanups:

The `ReduxWrappedLoggedOutLayout` component needs include also `MomentProvider`, otherwise one of the 2FA screens (waiting for mobile notification) will break, because it uses localized moment.

In `blocks/login`, there is a long `if-else-if-else` chain that determines `preHeader` and `postHeader` for various login variants: OAuth (Woo, Crowdsignal), Jetpack and Woo Connect, 2FA screens... I pushed a patch that makes sure that `ContinueAsUser` is shown only on the plain login page and we don't get screens like this:

<img width="421" alt="Screenshot 2019-08-28 at 13 59 41" src="https://user-images.githubusercontent.com/664258/63855053-ccd46c80-c99e-11e9-964e-b29e75d16129.png">

Separated a `validateUrl` function that's independent from the component. Similar to Redux action. Makes the `useEffect` code very short.

Using `null` as the initial state of the verified URL, meaning "we don't have a verified URL". Defaulting to `/` happens only when actually rendering. That simplified the logic a lot and the default URL is now used only at one place.

sgomes commented:
Thank you, @jsnajdr! These all look like great changes, and great catch on the 2FA screen! Thank you for all your help 🙇 

I added a comment, but it's really just a wish for the future more than anything requiring attention 🙂

jsnajdr reviewed (approved):
Ready to go if the e2e failures are verified as harmless :+1:

sgomes commented:
> Ready to go if the e2e failures are verified as harmless 👍

I'm not entirely sure they are. I believe that the E2E tests may be relying on the masterbar being available when opening `/log-in` as an already logged-in user.

@bsessions85 Could you confirm this? And if so, what do you think the right course of action to fix things would be, given that this PR removes the masterbar links in `/log-in`?

sgomes commented:
> I'm not entirely sure they are. I believe that the E2E tests may be relying on the logged-in masterbar being shown when opening `/log-in` as an already logged-in user.
> 
> @bsessions85 Could you confirm this? And if so, what do you think the right course of action to fix things would be, given that this PR removes the logged-in masterbar links in `/log-in`?

@Stojdza Perhaps you could help here instead?

Commit:
Check for continue link and click it if it shows up.

bsessions85 commented:
@sgomes My apologies for missing the ping on this one! I put in a small change that will check to see if the continue link is present and click it if it is. This will keep the tests moving. For some reason the e2e tests are still failing, but it isn't related to this PR. I have some changes in the works that should help that problem in the future.

sgomes commented:
> @sgomes My apologies for missing the ping on this one! I put in a small change that will check to see if the continue link is present and click it if it is. This will keep the tests moving. For some reason the e2e tests are still failing, but it isn't related to this PR. I have some changes in the works that should help that problem in the future.

No worries at all! Thank you so much for adding that in; I was looking for pointers but it looks like I got code instead! Even better 😄
